localstack/services/s3/provider_stream.py

```diff
@@ -0,0 +1,808 @@
+                with key_lock:
+                    while data := stream_value.read(4096):
+                        checksum.update(data)
+                    stream_value.seek(0)
+
+                S3_KEY_LOCKS.pop(stream_id, None)
+
+                calculated_checksum = base64.b64encode(checksum.digest()).decode()
+                dest_key_object.checksum_value = calculated_checksum
+            else:
+                dest_key_object.checksum_value = source_key_object.checksum_value
+            dest_key_object.checksum_algorithm = checksum_algorithm
+
+            if checksum_algorithm == ChecksumAlgorithm.CRC32C:
+                # TODO: the logic for rendering the template in moto is the following:
+                # if `CRC32` in `key.checksum_algorithm` which is valid for both CRC32 and CRC32C, and will render both
+                # remove the key if it's CRC32C.
+                response["CopyObjectResult"].pop("ChecksumCRC32", None)
+
+            response["CopyObjectResult"][
+                f"Checksum{checksum_algorithm.upper()}"
+            ] = dest_key_object.checksum_value  # noqa
+
+        self._notify(context)
+        return response
+
+    @handler("UploadPart", expand=False)
+    def upload_part(self, context: RequestContext, request: UploadPartRequest) -> UploadPartOutput:
+        bucket_name = request["Bucket"]
+        moto_backend = get_moto_s3_backend(context)
+        moto_bucket = get_bucket_from_moto(moto_backend, bucket_name)
+        if not (upload_id := request.get("UploadId")) in moto_bucket.multiparts:
+            raise NoSuchUpload(
+                "The specified upload does not exist. The upload ID may be invalid, or the upload may have been aborted or completed.",
+                UploadId=upload_id,
+            )
+        elif request.get("PartNumber", 0) < 1:
+            # TODO: find the right exception for this?
+            raise NoSuchUpload()
+
+        key = moto_backend.upload_part(
+            bucket_name, upload_id, request.get("PartNumber"), request.get("Body")
+        )
+        response = UploadPartOutput(ETag=key.etag)
+
+        if key.checksum_algorithm is not None:
+            response[f"Checksum{key.checksum_algorithm.upper()}"] = key.checksum_value
+
+        if key.encryption is not None:
+            response["ServerSideEncryption"] = key.encryption
+            if key.encryption == "aws:kms" and key.kms_key_id is not None:
+                response["SSEKMSKeyId"] = key.encryption
+
+        if key.encryption == "aws:kms" and key.bucket_key_enabled is not None:
+            response["BucketKeyEnabled"] = key.bucket_key_enabled
+
+        return response
+
+
+class ChecksumInvalid(s3_exceptions.S3ClientError):
+    code = 400
+
+    def __init__(self, algorithm: str):
+        super().__init__(
+            "InvalidRequest",
+            f"Value for x-amz-checksum-{algorithm.lower()} header is invalid.",
+        )
+
+
+class PartialStream(RawIOBase):
```

Comment:
The code below this line is getting less commented and more complex.
Since this is just the first iteration, and an opt-in provider, this seems to be good enough for now. A lot of this stuff will change, and a lot will be restructured. However, it would be great if parts like this could be commented a bit more in the upcoming iterations.

Reply:
Of course, as I'll fix comments for this review, I'll at least do minimal commenting of the code, I got a bit carried away here.